File: external/autotest/client/site_tests/video_YouTubeMseEme/video_YouTubeMseEme.py
# Copyright (c) 2013 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import logging, os

from autotest_lib.client.bin import test, utils
from autotest_lib.client.common_lib import error
from autotest_lib.client.common_lib.cros import chrome
from autotest_lib.client.cros import httpd


class video_YouTubeMseEme(test.test):
    """The main test class for MSE/EME.

    """


    version = 1

    PLAYER_PAGE = 'http://localhost:8000/files/video_YouTubeMseEme.html'
    TEST_JS = 'files/video_YouTubeMseEme.js'

    POLLING_TIME = 0.1


    def init(self, chrome, player_page):
        """Initialization function for this test class.

        @param chrome: An Autotest Chrome instance.
        @param player_page: Dummy HTML file to load.

        """
        self._testServer = httpd.HTTPListener(8000, docroot=self.bindir)
        self._testServer.run()

        self.tab = chrome.browser.tabs[0]
        self.tab.Navigate(player_page)
        self.tab.WaitForDocumentReadyStateToBeComplete()
        self.load_javascript(self.TEST_JS)


    def _check_event_happened(self, event_name):
        """A wrapper to check if an event in JS has fired.

        @param event_name: A string to denote the name of the event to check.
        """
        self.tab.WaitForJavaScriptExpression(
             'window.__eventReporter["%s"] === true;' % event_name, 5)


    def load_javascript(self, sub_path):
        """A wrapper to load a JS file into the current tab.

        @param sub_path: The relative path from the current .py file.

        """
        full_path = os.path.join(os.path.dirname(__file__), sub_path)
        with open(full_path) as f:
            js = f.read()
            self.tab.ExecuteJavaScript(js)
            logging.info('Loaded accompanying .js script.')


    def get_test_state(self, test_name, delay_time_sec=5):
        """A wrapper to check the state of a test in the accompanying JS.

        @param test_name: The name of the test that was ran.
        @param delay_time_sec: Time to wait before querying the test (float).
                This is to give the VM some time to schedule the next execution.

        @returns: A boolean value indicating the success of the test.

        """
        def _test_passed_condition(test_name):
            """An inner function to test if the test passed.

            @param test_name: The name of the test that was ran.

            @returns: A boolean indicating if the test passed.

            """
            return self.tab.EvaluateJavaScript(
                    'window.__testState["%s"];' % test_name)

        return utils.poll_for_condition(
                lambda: _test_passed_condition(test_name),
                timeout=delay_time_sec,
                desc=test_name)


    def test_media_source_presence(self):
        """Tests for the existence of the Media Source Extension.

        """
        self.assert_(
                self.tab.EvaluateJavaScript(
                        'window.WebKitMediaSource !== undefined'),
                msg='test_media_source_presence failed.')


    def test_attach_source(self):
        """Tests if attaching a the MediaSource to the video tag is successful.

        """
        self.tab.ExecuteJavaScript('window.__testAttach();')
        self._check_event_happened('sourceopen')


    def test_add_source_buffer(self):
        """Tests adding the source buffer to the MediaSource is successful.

        """
        self.tab.ExecuteJavaScript('window.__testAddSourceBuffer();')
        self.assert_(
                self.get_test_state('addSourceBuffer'),
                msg='test_add_source_buffer failed.')


    def test_add_supported_formats(self):
        """Tests adding supported formats to the MediaSource is successful.

        """
        self.tab.ExecuteJavaScript('window.__testAddSupportedFormats();')
        self.assert_(
                self.get_test_state('addSupportedFormats'),
                msg='test_add_supported_formats failed.')


    def test_add_source_buffer_exception(self):
        """Tests adding the source buffer to an uninitialized MediaSource.

        """
        self.tab.ExecuteJavaScript(
                'window.__testAddSourceBufferException();')
        self.assert_(
                self.get_test_state('addSourceBufferException'),
                msg='test_add_source_buffer_exception failed.')


    def test_initial_video_state(self):
        """Tests the initial states of the video object.

        """
        self.tab.ExecuteJavaScript('window.__testInitialVideoState();')
        self.assert_(
                self.get_test_state('initialVideoState'),
                msg='test_initial_video_state failed.')


    def test_initial_media_source_state(self):
        """Tests the initial states of the MediaSource object.

        """
        self.tab.ExecuteJavaScript('window.__testInitialMSState();')
        self.assert_(
                self.get_test_state('initialMSState'),
                msg='test_initial_media_source_state failed.')


    def test_append(self):
        """Tests appending audio and video streams.

        """
        self.tab.ExecuteJavaScript('window.__testAppend_audio();')
        self.assert_(
                self.get_test_state('append_audio'),
                msg='test_append failed on audio append.')

        self.tab.ExecuteJavaScript('window.__testAppend_video();')
        self.assert_(
                self.get_test_state('append_video'),
                msg='test_append failed on video append.')


    def test_append_abort(self):
        """Tests appending followed by aborting audio and video streams.

        """
        self.tab.ExecuteJavaScript('window.__testAppendAbort_audio();')
        self.assert_(
                self.get_test_state('appendAbort_audio'),
                msg='test_append_abort failed on audio.')

        self.tab.ExecuteJavaScript('window.__testAppendAbort_video();')
        self.assert_(
                self.get_test_state('appendAbort_video'),
                msg='test_append_abort failed on video.')


    def test_append_timestamp_offset(self):
        """Tests appending with a timestamp offset.

        """
        self.tab.ExecuteJavaScript(
                'window.__testAppendTimestampOffset_audio();')
        self.assert_(
                self.get_test_state('appendTimestampOffset_audio'),
                msg='test_append_timestamp_offset failed on audio.')

        self.tab.ExecuteJavaScript(
                'window.__testAppendTimestampOffset_video();')
        self.assert_(
                self.get_test_state('appendTimestampOffset_video'),
                msg='test_append_timestamp_offset failed on video.')


    def test_duration(self):
        """Tests if it's possible to set the duration on the Media Source.

        """
        self.tab.ExecuteJavaScript('window.__testDuration();')
        self.assert_(
                self.get_test_state('duration'),
                msg='test_duration failed.')


    def test_duration_after_append(self):
        """Tests if the duration changes after appending to the source buffer.

        """
        self.tab.ExecuteJavaScript('window.__testDurationAfterAppend_audio();')
        self.assert_(
                self.get_test_state('durationAfterAppend_audio'),
                msg='test_duration_after_append failed on audio.')

        self.tab.ExecuteJavaScript('window.__testDurationAfterAppend_video();')
        self.assert_(
                self.get_test_state('durationAfterAppend_video'),
                msg='test_duration_after_append failed on video.')


    def test_source_remove(self):
        """Tests if removing the source buffers works correctly.

        """
        self.tab.ExecuteJavaScript('window.__testSourceRemove();')
        self.assert_(
                self.get_test_state('sourceRemove'),
                msg='test_source_remove failed.')


    def test_can_play_webm(self):
        """Tests if it's possible to play WebM content.

        """
        self.assert_(
                self.tab.EvaluateJavaScript('window.__testCanPlayWebM();'),
                msg='test_can_play_webm failed.')


    def test_can_play_clear_key(self):
        """Tests if it's possible to play ClearKey content.

        """
        self.assert_(
                self.tab.EvaluateJavaScript('window.__testCanPlayClearKey();'),
                msg='test_can_play_clear_key failed.')


    def test_can_not_play_play_ready(self):
        """Tests if it's impossible to play PlayReady.

        """
        self.assert_(
                self.tab.EvaluateJavaScript(
                        'window.__testCanNotPlayPlayReady();'),
                msg='test_can_not_play_play_ready failed.')


    def test_can_play_widevine(self):
        """Tests if it's possible to play Widevine content.

        """
        self.assert_(
                self.tab.EvaluateJavaScript('window.__testCanPlayWidevine();'),
                msg='test_can_play_widevine failed.')


    def run_once(self, subtest_name):
        with chrome.Chrome() as cr:
            self.init(cr, self.PLAYER_PAGE)

            try:
                # The control file passes in a test name, which is the name of
                #  the test to run, prepended with 'test_'.
                function_to_call = getattr(self, 'test_' + subtest_name)
                function_to_call()
            except AttributeError:
                # Just in case the input test name was mistyped in the control
                #  file.
                raise error.TestFail('No function named: test_' + subtest_name)

File: external/autotest/client/site_tests/video_YouTubeMseEme/files/video_YouTubeMseEme.js
// Copyright (c) 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
"use strict";

(function() {
  window.__eventReporter = {};
  window.__testState = {};

  var logger = new Logger();
  window.__logger = logger;

  var video_format = 'video/mp4; codecs="avc1.640028"';
  var audio_format = 'audio/mp4; codecs="mp4a.40.2"';

  var audio1MB = createAudioDef(
      'http://localhost:8000/files/car-audio-1MB-trunc.mp4', 1048576, 65.875);
  var video1MB = createVideoDef(
      'http://localhost:8000/files/test-video-1MB.mp4', 1031034, 1.04);

  // Utility functions.
  function Logger() {
    var logs = [];

    this.log = function(log_string) {
      logs.push(log_string);
    };

    this.toString = function() {
      var output = '';
      for (var i in logs)
        output += logs[i];

      return output;
    }
  }

  function createAudioDef(src, size, duration) {
    return {
      type: 'audio',
      format: audio_format,
      size: size,
      src: src,
      duration: duration,
      bps: Math.floor(size / duration)
    };
  }

  function createVideoDef(src, size, duration) {
    return {
      type: 'video',
      format: video_format,
      size: size,
      src: src,
      duration: duration,
      bps: Math.floor(size / duration)
    };
  }

  function createMediaSource() {
    if (typeof MediaSource !== 'undefined')
      return new MediaSource();
    else
      return new WebKitMediaSource();
  }

  function createVideo() {
    return document.createElement('video');
  }

  function setupVideoAndMs(onSourceopen) {
    var temp_video = createVideo();
    var ms = createMediaSource();
    ms.addEventListener('webkitsourceopen', onSourceopen);
    var ms_url = window.URL.createObjectURL(ms);
    temp_video.src = ms_url;
    return {
      'video': temp_video,
      'ms': ms
    };
  }

  function XHRWrapper(file, onLoad, onError, start, length) {
    self = this;

    self.file = file;
    self.onLoad = onLoad;
    self.onError = onError;
    self.start = start;
    self.length = length;

    this.getResponseData = function() {
      var result = new Uint8Array(this.xhr.response);
      if (start != null) {
        return result.subarray(start, start + length);
      }
      return result;
    };

    this.abort = function() {
      this.xhr.abort();
    };

    this.send = function() {
      this.xhr.send();
    };

    this.xhr = new XMLHttpRequest();
    this.xhr.open('GET', file, true);

    this.xhr.addEventListener('load', function(e) {
      self.onLoad(e);
    });

    this.xhr.addEventListener('error', function(e) {
      logger.log('XHR errored.');
      self.onError(e);
    });

    this.xhr.addEventListener('timeout', function(e) {
      logger.log('XHR timed out.');
      self.onError(e);
    });

    this.xhr.responseType = 'arraybuffer';
    if (length != null) {
      start = start || 0;
      this.xhr.setRequestHeader(
        'Range',
        'bytes=' + start + '-' + (start + length - 1)
      );
    }
  }

  function approxEq(a, b) {
    return Math.abs(a - b) < 0.5;
  };

  // MSE tests.
  window.__testAttach = function() {
    var ms = createMediaSource();
    ms.addEventListener('webkitsourceopen', function() {
      window.__eventReporter['sourceopen'] = true;
    });

    var video = document.getElementById('main_player');
    video.src = window.URL.createObjectURL(ms);
    video.load();
  };

  window.__testAddSourceBuffer = function() {
    var vm = setupVideoAndMs(function() {
      try {
        var return_value = true;
        return_value &= vm.ms.sourceBuffers.length === 0;
        vm.ms.addSourceBuffer(audio_format);
        return_value &= vm.ms.sourceBuffers.length === 1;
        vm.ms.addSourceBuffer(video_format);
        return_value &= vm.ms.sourceBuffers.length === 2;

        window.__testState['addSourceBuffer'] = !!return_value;
      }
      catch (e) {
        window.__testState['addSourceBuffer'] = false;
      }
    });
  };

  window.__testAddSupportedFormats = function() {
    var formats = [
      audio_format,
      video_format,
    ];

    var vm = setupVideoAndMs(function() {
      for (var i = 0; i < formats.length; ++i) {
        try {
          vm.ms.addSourceBuffer(formats[i]);
        } catch (e) {
          window.__testState['addSupportedFormats'] = false;
          return;
        }
      }
      window.__testState['addSupportedFormats'] = true;
    });
  };

  window.__testAddSourceBufferException = function() {
    var vm = setupVideoAndMs(function() {
      try {
        vm.ms.addSourceBuffer('^^^');
        window.__testState['addSourceBufferException'] = false;
        return;
      }
      catch (e) {
        if (e.code !== DOMException.NOT_SUPPORTED_ERR) {
          window.__testState['addSourceBufferException'] = false;
          return;
        }
      }

      try {
        var temp_media_source = new WebKitMediaSource();
        temp_media_source.addSourceBuffer(audio_format);
        window.__testState['addSourceBufferException'] = false;
        return;
      }
      catch (e) {
        if (e.code !== DOMException.INVALID_STATE_ERR) {
          window.__testState['addSourceBufferException'] = false;
          return;
        }
      }
      window.__testState['addSourceBufferException'] = true;
    });
  };

  window.__testInitialVideoState = function() {
    var temp_video = createVideo();

    var test_result = isNaN(temp_video.duration);
    test_result &= temp_video.videoWidth === 0;
    test_result &= temp_video.videoHeight === 0;
    test_result &= temp_video.readyState === HTMLMediaElement.HAVE_NOTHING;
    test_result &= temp_video.src === '';
    test_result &= temp_video.currentSrc === '';

    window.__testState['initialVideoState'] = !!test_result;
  };

  window.__testInitialMSState = function() {
    var vm = setupVideoAndMs(
      function() {
        var test_result = true;
        test_result = test_result && isNaN(vm.ms.duration);
        test_result = test_result && vm.ms.readyState === 'open';
        window.__testState['initialMSState'] = test_result;
    });
  };

  function appendTestTemplate(
    test_name, media, test_func, abort, start, length, offset) {
    var vm = setupVideoAndMs(function() {
      var sb = vm.ms.addSourceBuffer(media.format);
      var xhr = new XHRWrapper(
        media.src,
        function(e) {
          var response_data = xhr.getResponseData();

          if (offset != null)
            sb.timestampOffset = offset;

          sb.append(response_data);

          if (abort != null) {
            sb.abort();
            sb.append(response_data);
          }

          var test_result = test_func(sb, media);
          window.__testState[test_name] = test_result;
        },
        function(e) {
          window.__testState[test_name] = false;
        }, start, length);

      xhr.send();
    });
  }

  function appendInnerTest(sb, media) {
    return sb.buffered.length === 1 && sb.buffered.start(0) === 0 &&
      approxEq(sb.buffered.end(0), media.duration);
  }

  window.__testAppend_audio = function() {
    appendTestTemplate('append_audio', audio1MB, appendInnerTest);
  };

  window.__testAppend_video = function() {
    appendTestTemplate('append_video', video1MB, appendInnerTest);
  };

  function appendAbortInnerTest(sb, media) {
    return sb.buffered.length === 1 && sb.buffered.start(0) === 0 &&
      sb.buffered.end(0) > 0;
  }

  window.__testAppendAbort_audio = function() {
    appendTestTemplate(
      'appendAbort_audio', audio1MB, appendAbortInnerTest, true, 0, 200000);
  };

  window.__testAppendAbort_video = function() {
    appendTestTemplate(
      'appendAbort_video', video1MB, appendAbortInnerTest, true, 0, 200000);
  };

  var TIMESTAMP_BUFFERED_OFFSET = 5;
  function appendTimestampOffsetTest(sb, media) {

    return sb.buffered.length === 1 &&
      sb.buffered.start(0) === TIMESTAMP_BUFFERED_OFFSET &&
      approxEq(sb.buffered.end(0), media.duration + TIMESTAMP_BUFFERED_OFFSET);
  }

  window.__testAppendTimestampOffset_audio = function() {
    appendTestTemplate(
      'appendTimestampOffset_audio', audio1MB, appendTimestampOffsetTest,
      null, null, null, TIMESTAMP_BUFFERED_OFFSET);
  };

  window.__testAppendTimestampOffset_video = function() {
    appendTestTemplate(
      'appendTimestampOffset_video', video1MB, appendTimestampOffsetTest,
      null, null, null, TIMESTAMP_BUFFERED_OFFSET);
  };

  window.__testDuration = function() {
    var vm = setupVideoAndMs(
      function() {
        var DURATION_TIME = 10;
        vm.ms.duration = DURATION_TIME;
        window.setTimeout(function() {
            window.__testState['duration'] = vm.ms.duration === DURATION_TIME;
        }, 20);
    });
  };

  function testDurationAfterAppend(test_name, media) {
    var vm = setupVideoAndMs(function() {
      var sb = vm.ms.addSourceBuffer(media.format);

      function onDurationChange() {
        window.__testState[test_name] = approxEq(
          vm.ms.duration, sb.buffered.end(0));
      }

      var xhr = new XHRWrapper(media.src, function() {
          var response_data = xhr.getResponseData();
          sb.append(response_data);
          sb.abort();
          vm.ms.duration = sb.buffered.end(0) / 2;
          vm.video.addEventListener('durationchange', onDurationChange);
          sb.append(response_data);
        });

      xhr.send();
    });
  };

  window.__testDurationAfterAppend_audio = function() {
    testDurationAfterAppend('durationAfterAppend_audio', audio1MB);
  };

  window.__testDurationAfterAppend_video = function() {
    testDurationAfterAppend('durationAfterAppend_video', video1MB);
  };

  window.__testSourceRemove = function() {
    var vm = setupVideoAndMs(
      function() {
        var sbAudio = vm.ms.addSourceBuffer(audio_format);
        var result = vm.ms.sourceBuffers.length === 1;
        vm.ms.removeSourceBuffer(sbAudio);
        result &= vm.ms.sourceBuffers.length === 0;

        sbAudio = vm.ms.addSourceBuffer(audio_format);
        result &= vm.ms.sourceBuffers.length === 1;
        for (var i = 0; i < 10; ++i) {
          var sbVideo = vm.ms.addSourceBuffer(video_format);
          result &= vm.ms.sourceBuffers.length === 2;
          vm.ms.removeSourceBuffer(sbVideo);
          result &= vm.ms.sourceBuffers.length === 1;
        }

        vm.ms.removeSourceBuffer(sbAudio);
        result &= vm.ms.sourceBuffers.length === 0;

        window.__testState['sourceRemove'] = !!result;
    });
  };

  // EME tests.
  window.__testCanPlayWebM = function() {
    var tempVideo = createVideo();
    return tempVideo.canPlayType(
        'video/webm; codecs="vp8,vorbis"') === 'probably' &&
      tempVideo.canPlayType(
        'audio/webm; codecs="vorbis"') === 'probably';
  };

  window.__testCanPlayClearKey = function() {
    var tempVideo = createVideo();
    return tempVideo.canPlayType(
        'video/mp4; codecs="avc1.640028"',
        'webkit-org.w3.clearkey') === 'probably' &&
      tempVideo.canPlayType(
        'audio/mp4; codecs="mp4a.40.2"',
        'webkit-org.w3.clearkey') === 'probably';
  };

  window.__testCanNotPlayPlayReady = function() {
    var tempVideo = createVideo();
    return tempVideo.canPlayType(
        'video/mp4; codecs="avc1.640028"',
        'com.youtube.playready') !== 'probably' &&
      tempVideo.canPlayType(
        'audio/mp4; codecs="mp4a.40.2"',
        'com.youtube.playready') !== 'probably';
  };

  window.__testCanPlayWidevine = function() {
    function createWidevineTest(mediaType) {
      var tempVideo = createVideo();

      return function(codecs, keySystem, criteria) {
        var codecString = mediaType;
        if (codecs != null)
          codecString += '; codecs="' + codecs + '"';

        var testResult = tempVideo.canPlayType(codecString, keySystem);

        if (criteria === null)
          return testResult === 'probably' || testResult === 'maybe';
        else if (typeof(criteria) === 'string')
          return testResult === criteria;
        else if (criteria.length) {
          var checks = false;
          for (var i in criteria)
            checks |= testResult === criteria[i];
          return !!checks;
        }
        return false;
      }
    }

    var audioTest = createWidevineTest('audio/webm');
    var videoTest = createWidevineTest('video/webm');

    var result = true;

    // Supported video formats.
    result &= videoTest(null, 'com.widevine.alpha', 'maybe');
    result &= videoTest(null, 'com.widevine', 'maybe');
    result &= videoTest('vp8', 'com.widevine.alpha', 'probably');
    result &= videoTest('vp8', 'com.widevine', 'probably');
    result &= videoTest('vp8.0', 'com.widevine.alpha', 'probably');
    result &= videoTest('vp8.0', 'com.widevine', 'probably');
    result &= videoTest('vorbis', 'com.widevine.alpha', 'probably');
    result &= videoTest('vorbis', 'com.widevine', 'probably');
    result &= videoTest('vp8,vp8.0,vorbis', 'com.widevine.alpha', 'probably');
    result &= videoTest('vp8,vp8.0,vorbis', 'com.widevine', 'probably');

    // Supported audio formats.
    result &= audioTest(null, 'com.widevine.alpha', 'maybe');
    result &= audioTest(null, 'com.widevine', 'maybe');
    result &= audioTest('vorbis', 'com.widevine.alpha', 'probably');
    result &= audioTest('vorbis', 'com.widevine', 'probably');

    // Unsupported video formats.
    result &= videoTest('codecs="vp8"', 'com.widevine.', '');
    result &= videoTest('codecs="vp8"', 'com.widevine.foo', '');
    result &= videoTest('codecs="vp8"', 'com.widevine.alpha.', '');
    result &= videoTest('codecs="vp8"', 'com.widevine.alpha.foo', '');
    result &= videoTest('codecs="vp8"', 'com.widevine.alph', '');
    result &= videoTest('codecs="vp8"', 'com.widevine.alphb', '');
    result &= videoTest('codecs="vp8"', 'com.widevine.alphaa', '');
    result &= videoTest('codecs="avc1.640028"', 'com.widevine.alpha', '');
    result &= videoTest('codecs="mp4a"', 'com.widevine.alpha', '');

    // Unsported audio formats.
    result &= audioTest('codecs="vp8"', 'com.widevine', '');
    result &= audioTest('codecs="vp8,vorbis"', 'com.widevine.alpha', '');
    result &= audioTest('codecs="vp8,vorbis"', 'com.widevine.alpha', '');

    return result;
  };
})();
